packages/common/src/TransactionUtils.ts

```diff
@@ -101,35 +130,52 @@ export const runTransaction = async ({
     const maximumGasPriceMultiple = 2 * 3;
     // Pre-London transactions require `gasPrice`, London transactions require `maxFeePerGas` and `maxPriorityFeePerGas`
 
-    let receipt;
+    let receipt: TransactionReceipt | PromiEvent<TransactionReceipt>;
+    let transactionHash: string;
 
     // If the config contains maxPriorityFeePerGas then this is a London transaction. In this case, simply use the
     // provided config settings but double the maxFeePerGas to ensure the transaction is included, even if the base fee
     // spikes up. The difference between the realized base fee and maxFeePerGas is refunded in a London transaction.
     if (transactionConfig.maxFeePerGas && transactionConfig.maxPriorityFeePerGas) {
-      receipt = await transaction.send({
-        ...transactionConfig,
-        maxFeePerGas: parseInt(transactionConfig.maxFeePerGas.toString()) * 2,
-        type: "0x2",
-      } as SendOptions);
+      // If waitForMine is set (default) then code blocks until the transaction is mined and a receipt is returned.
+      if (waitForMine) {
+        receipt = ((await transaction.send({
+          ...transactionConfig,
+          maxFeePerGas: parseInt(transactionConfig.maxFeePerGas.toString()) * 2,
+          type: "0x2",
+        } as SendOptions)) as unknown) as TransactionReceipt;
+        transactionHash = receipt.transactionHash;
+      }
+      // Else, waitForMine is false and we return the transaction hash immediately as soon as it is included in the
+      // mempool. Receipt is a promise of the pending transaction that can be awaited later to ensure block inclusion.
+      else {
+        receipt = (transaction.send({
+          ...transactionConfig,
+          maxFeePerGas: parseInt(transactionConfig.maxFeePerGas.toString()) * 2,
+          type: "0x2",
+        } as SendOptions) as unknown) as PromiEvent<TransactionReceipt>;
+        transactionHash = await new Promise((resolve) => {
+          (receipt as any).on("transactionHash", (transactionHash: any) => resolve(transactionHash));
```

Comment:
Does PromiEvent not have the `on` method?

Reply:
no :(